RustLeanModels/RustString.lean

```diff
@@ -0,0 +1,4914 @@
+  rename_i hs1 ts1; simp only [byteSize_cons] at g1
+  have g2:= prefix_cons g1.left ; rw[g2.left] at g1 ;
+  simp only [not_exists, not_and] at g; have g:=g ts1 g2.right; omega
+  --mpr
+  simp only [not_exists, not_and]; intro g
+  split; have g:= g []; simp only [_root_.nil_prefix, byteSize, byteSize_aux, forall_true_left] at g; omega
+  split; simp only; apply ind.mpr
+  by_contra; rename_i g1; obtain ⟨s1, g1⟩ := g1
+  have : List.IsPrefix (h :: s1) (h :: t) :=by apply prefix_cons_mpr g1.left
+  have g:= g (h::s1) this; simp only [byteSize_cons] at g; omega
+
+theorem split_at_none_sound : split_at s i = none ↔ ¬ ∃ s1, List.IsPrefix s1 s ∧ byteSize s1 = i :=by
+  unfold split_at; exact split_at_aux_none_sound
+
+
+theorem split_at_some_sound : split_at s i= some (s1, s2) ↔ byteSize s1 = i ∧ s = s1 ++ s2 :=by
+  induction s generalizing i s1 s2
+  unfold split_at split_at_aux ; split
+  constructor; simp only [Option.some.injEq, Prod.mk.injEq, nil_eq_append, and_imp]
+  intro g1 g2; rename_i gi; rw[← g1, gi]; simp only [g2, and_self, and_true]; rw[byteSize_aux_para1_elim]
+  rw[← g2]; simp only [byteSize_aux]
+  simp only [nil_eq_append, Option.some.injEq, Prod.mk.injEq, and_imp]
+  intro _ g2 g3; rw[g2,g3]; simp only [and_self]
+  constructor; simp only [nil_eq_append, IsEmpty.forall_iff]
+  simp only [nil_eq_append, and_imp, imp_false]; intro g1 g2
+  rw[g2] at g1; simp only [byteSize, byteSize_aux] at g1; omega
+  rename_i h t ind
+  constructor
+  intro g; unfold split_at split_at_aux at g
+  split at g; simp only [Option.some.injEq, Prod.mk.injEq] at g; rename_i gi
+  rw[gi, ← g.left]; simp only [byteSize, byteSize_aux, g.right, List.nil_append, and_self]
+  split at g; contradiction; simp only [List.nil_append] at g
+  have g:= split_at_aux_para1_elim.mp g; obtain ⟨t1, g⟩ := g
+  have g1:= ind.mp g.right
+  simp only [g, singleton_append, byteSize_cons, g1, cons_append, and_true];
+  simp only [List.nil_append, and_true]; omega
+  --mpr
+  intro g ; unfold split_at split_at_aux; split ; rename_i gi
+  rw[gi] at g; have gs1:= nil_of_byteSize_zero g.left ; rw[g.right, gs1]; simp only [List.nil_append]
+  split; cases s1; simp only [byteSize, byteSize_aux, List.nil_append] at g; omega
+  rename_i hs1 ts1; simp only [byteSize_cons, cons_append, cons.injEq] at g;
+  rename_i gi; rw [g.right.left] at gi; omega
+  simp only [List.nil_append]; rw[split_at_aux_para1_elim]
+  cases s1; simp only [byteSize, byteSize_aux, List.nil_append] at g; omega
+  rename_i hs1 ts1; simp only [byteSize_cons, cons_append, cons.injEq] at g;
+  use ts1; simp only [g, singleton_append, true_and]; rw[← g.right.right]
+  apply ind.mpr; constructor; omega; exact g.right.right
+
+
+lemma split_at_zero: split_at s 0 = some ([], s) := by
+  unfold split_at split_at_aux; simp only [↓reduceIte]
+  split; simp only; simp only
+
+lemma split_at_is_char_boundary_none: ¬ is_char_boundary s i ↔  split_at s i = none :=by
```

Comment:
Would it make sense to name the lemma `split_at_non_char_boundary_none` instead?